performance.md

```diff
@@ -0,0 +1,227 @@
+
+``` r
+lazy_query = lf_test$
+  sort(pl$col("country"))$
+  filter(
+    pl$col("country")$is_in(pl$lit(c("United Kingdom", "Japan", "Vietnam")))
+  )
+
+lazy_query
+```
+
+    ## [1] "polars LazyFrame naive plan: (run ldf$describe_optimized_plan() to see the optimized plan)"
+    ##   FILTER col("country").is_in([Series]) FROM
+    ##     SORT BY [col("country")]
+    ##       CSV SCAN data/test.csv
+    ##       PROJECT */3 COLUMNS
+
+However, this doesn’t do anything to the data until we call `collect()`
+at the end. We can now compare the two approaches:
+
+``` r
+bench::mark(
+  eager = df_test$
+    sort(pl$col("country"))$
+    filter(
+      pl$col("country")$is_in(pl$lit(c("United Kingdom", "Japan", "Vietnam")))
+    ),
+  lazy = lazy_query$collect(),
+  iterations = 10
+)
+```
+
+    ## # A tibble: 2 × 6
+    ##   expression      min   median `itr/sec` mem_alloc `gc/sec`
+    ##   <bch:expr> <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
+    ## 1 eager         1.13s    1.18s     0.839    7.94KB        0
+    ## 2 lazy       524.08ms  557.8ms     1.78       640B        0
+
+On this very simple query, using lazy execution instead of eager
+execution lead to a 1.7-2.2x decrease in time.
+
+So what happened? Under the hood, `polars` reorganized the query so that
+it filters the data first, and then sorts the remaining data. This can
+be seen by comparing the original query (`describe_plan()`) and the
+optimized query (`describe_optimized_plan()`):
+
+``` r
+lazy_query$describe_plan()
+```
+
+    ##   FILTER col("country").is_in([Series]) FROM
+    ##     SORT BY [col("country")]
+    ##       CSV SCAN data/test.csv
+    ##       PROJECT */3 COLUMNS
+
+``` r
+lazy_query$describe_optimized_plan()
+```
+
+    ##   SORT BY [col("country")]
+    ##     CSV SCAN data/test.csv
+    ##     PROJECT */3 COLUMNS
+    ##     SELECTION: col("country").is_in([Series])
+
+Note that the queries must be read from bottom to top, i.e the optimized
+query is “select the dataset where the column ‘country’ matches these
+values, then sort the data by the values of ‘country’”.
+
+# Use `polars` functions
+
+`polars` comes with a large number of built-in, optimized, basic
+functions that should cover most aspects of data wrangling. These
+functions are designed to be very memory efficient. Therefore, using R
+functions or converting data back and forth between `polars` and R is
+discouraged as it can lead to a large decrease in efficiency.
+
+Let’s use the test data from the previous section and let’s say that we
+only want to check whether each country contains “na”. This can be done
+in (at least) two ways: with the built-in function `contains()` and with
+the base R function `grepl()`. However, using the built-in function is
+much faster:
+
+``` r
+bench::mark(
+  contains = df_test$with_columns(
+    pl$col("country")$str$contains("na")
+  ),
+  grepl = df_test$with_column(
+    pl$col("country")$map(\(s) { # map with a R function
```

Comment:
I think it's important to make that point, but also expect that a number of readers will find this comparison unfair. Isn't it like applying `grepl()` to every element individually instead of the vector? That's against the R mantra of vectorizing operations...

Reply:
I was surprised Series can be passed directly to grepl, I guess that is all the sweet S3 methods at play :)

Reply:
map has a very slow cousing called apply which should never be used except for in group_by calls.
If no group_by apply will iterate and convert every element individually

```R
grepl_scalarized = df_test$limit(1e6)$with_column( #10 x fewer
    pl$col("country")$apply(\(str) { # map with a R function
      grepl("na", str)
    }, return_type = pl$Boolean)
  )
```

in a group_by context apply iterate over the group, and perform a map on each group. If using map in a group_by context, it ignores the grouping and transform the raw column.

Reply:
btw the map could be written with explicit conversions, depending on style


```R
  grepl_explicit = df_test$with_column(
    pl$col("country")$map(\(s) { # map with a R function
      pl$Series(grepl("na", s$to_vector())) #explicit conversion
    })
  ),

```

Reply:
nice, thanks for the details